accounts/views.py

```diff
@@ -1,3 +1,291 @@
+            return reverse("accounts:2fa-login")
+        elif user.is_researcher or user.is_staff:
+            messages.warning(
+                self.request,
+                "If you're a researcher or Lookit staff, you'll want to set up "
+                "2FA with us. Please complete the Two-Factor Auth setup below "
+                "and you'll be on your way!",
+            )
+            return reverse("accounts:2fa-setup")
+        else:
+            return super().get_success_url()
+
+
+class TwoFactorAuthLoginView(UserPassesTestMixin, LoginView):
+    """Semi-optional two-factor authentication login.
+
+    Researchers must have 2FA activated and verified prior to viewing
+    /exp/ pages. Participants are precluded from visiting this page.
+    """
+
+    form_class = TOTPCheckForm
+
+    def user_is_researcher_or_staff(self):
+        return getattr(self.request.user, "is_researcher") or getattr(
+            self.request.user, "is_staff"
+        )
+
+    test_func = user_is_researcher_or_staff
+
+    def form_valid(self, form):
+        """Override base functionality to skip auth part.
+
+        Since OTP was already checked during Form cleaning process, we can just
+        redirect here.
+        """
+        self.request.session[TWO_FACTOR_AUTH_SESSION_KEY] = True
+        return HttpResponseRedirect(self.get_success_url())
+
+    def get_redirect_url(self) -> str:
+        """Have a good default for experimenters - the study list."""
+        return super().get_redirect_url() or reverse("exp:study-list")
+
+
+class ResearcherRegistrationView(generic.CreateView):
+    template_name = "accounts/researcher-registration.html"
+    model = User
+    form_class = forms.ResearcherRegistrationForm
+
+    def form_valid(self, form):
+        """If the registration process went well, log the user in."""
+        # UserRegistrationForm.is_valid() should do proper authentication
+        resp = super().form_valid(form)
+
+        # We expect user to be loaded by `ModelFormMixin.form_valid`
+        user: User = getattr(self, "object")
+
+        # Following with what django.auth.views.LoginView does here.
+        login(
+            self.request,
+            user,
+            backend="accounts.backends.TwoFactorAuthenticationBackend",
+        )
+        messages.success(self.request, "Researcher account created.")
+        return resp
+
+    def get_success_url(self) -> str:
+        """Researchers to 2FA setup, Participants go to demographic data update.
+
+        Leverage the fact that `ModelFormMixin.form_valid` sets `self.object = form.save()`
+        prior to calling the super() method (`FormMixin.form_valid`), which actually does
+        the work of constructing the HttpRedirectResponse for us.
+        """
+        user: User = getattr(self, "object")
+        view_target = (
+            "accounts:2fa-setup"
+            if user.is_researcher
+            else "web:demographic-data-update"
```

Comment:
ooh actually we shouldn't be sending participants to demographic data update unless they have no existing demographic data. Otherwise it'll be annoying to get sent to the demographic form every time they log in.

Reply:
Well actually this is `ResearcherRegistrationView`, not `ParticipantSignupView`, so I'm not really sure why I had that check in the first place... gonna delete it

Reply:
Also if they're registering and logging in as their user account for the first time, they'll never have demographic data